Calculate pages checksum in the zenith WAL redo loop

ololobus opened the pull request.

lubennikovaav committed:
[smgr_api] [community] smgr_api.patch

Make smgr API pluggable. Add smgr_hook that can be used to define custom smgrs.
Remove smgrsw[] array and smgr_sw selector. Instead, smgropen() loads
f_smgr implementation using smgr_hook.

Also add smgr_init_hook and smgr_shutdown_hook.
And a lot of mechanical changes in smgr.c functions.

This patch is proposed to community: https://commitfest.postgresql.org/33/3216/

Author: anastasia <lubennikovaav@gmail.com>

lubennikovaav committed:
[contrib/zenith] contrib_zenith.patch

Add contrib/zenith that handles interaction with remote pagestore.
To use it add 'shared_preload_library = zenith' to postgresql.conf.

It adds a protocol for network communications - see libpagestore.c;
and implements smgr API.

Also it adds several custom GUC variables:
- zenith.page_server_connstring
- zenith.callmemaybe_connstring
- zenith.zenith_timeline
- zenith.wal_redo

Authors:
Stas Kelvich <stanconn@gmail.com>
Konstantin Knizhnik <knizhnik@garret.ru>
Heikki Linnakangas <heikki.linnakangas@iki.fi>

lubennikovaav committed:
[walredo] zenith_wal_redo.patch

Add WAL redo helper for zenith - alternative postgres operation mode to replay wal by pageserver request.

To start postgres in wal-redo mode, run postgres with --wal-redo option
It requires zenith shared library and zenith.wal_redo

Author: Heikki Linnakangas <heikki.linnakangas@iki.fi>

lubennikovaav committed:
lastWrittenPageLSN.patch

Save lastWrittenPageLSN in XLogCtlData to know what pages to request from remote pageserver.

Authors:
Konstantin Knizhnik <knizhnik@garret.ru>
Heikki Linnakangas <heikki.linnakangas@iki.fi>

lubennikovaav committed:
handle_eviction_of_non_wal_logged_pages.patch

Some operations in PostgreSQL are not WAL-logged at all (i.e. hint bits)
or delay wal-logging till the end of operation (i.e. index build).
So if such page is evicted, we will lose the update.

To fix it, we introduce PD_WAL_LOGGED bit to track whether the page was wal-logged.
If the page is evicted before it has been wal-logged, then zenith smgr creates FPI for it.

Authors:
Konstantin Knizhnik <knizhnik@garret.ru>
anastasia <lubennikovaav@gmail.com>

lubennikovaav committed:
[walproposer] wal_proposer.patch

Add WalProposer background worker to broadcast WAL stream to Zenith WAL acceptors

Author: Konstantin Knizhnik <knizhnik@garret.ru>

lubennikovaav committed:
persist_unlogged_tables.patch

Ignore unlogged table qualifier. Add respective changes to regression test outputs.

Author: Konstantin Knizhnik <knizhnik@garret.ru>

lubennikovaav committed:
fix_pg_table_size.patch

Request relation size via smgr function, not just stat(filepath).

lubennikovaav committed:
[walredo] fix_gin_redo.patch

Author: Konstantin Knizhnik <knizhnik@garret.ru>

lubennikovaav committed:
[walredo] fix_brin_redo.patch

Author: Konstantin Knizhnik <knizhnik@garret.ru>

lubennikovaav committed:
speculative_records_workaround.patch

lubennikovaav committed:
wallog_t_ctid.patch

lubennikovaav committed:
vacuumlazy_debug_stub.patch

lubennikovaav committed:
[test] zenith_test_evict.patch

lubennikovaav committed:
fix_sequence_wallogging.patch

lubennikovaav committed:
Bring back change that got lost in refactoring. silence ReadBuffer_common error. TODO: add a comment, why this is fine for zenith.

hlinnaka committed:
Fix GetPage requests right after replaying CREATE DATABASE

In the test_createdb test, we created a new database, and created a new
branch after that. I was seeing the test fail with:

    PANIC:  could not open critical system index 2662

The WAL contained records like this:

    rmgr: XLOG        len (rec/tot):     49/  8241, tx:          0, lsn: 0/0163E8F0, prev 0/0163C8A0, desc: FPI , blkref #0: rel 1663/12985/1249 fork fsm blk 1 FPW
    rmgr: XLOG        len (rec/tot):     49/  8241, tx:          0, lsn: 0/01640940, prev 0/0163E8F0, desc: FPI , blkref #0: rel 1663/12985/1249 fork fsm blk 2 FPW
    rmgr: Standby     len (rec/tot):     54/    54, tx:          0, lsn: 0/01642990, prev 0/01640940, desc: RUNNING_XACTS nextXid 541 latestCompletedXid 539 oldestRunningXid 540; 1 xacts: 540
    rmgr: XLOG        len (rec/tot):    114/   114, tx:          0, lsn: 0/016429C8, prev 0/01642990, desc: CHECKPOINT_ONLINE redo 0/163C8A0; tli 1; prev tli 1; fpw true; xid 0:541; oid 24576; multi 1; offset 0; oldest xid 532 in DB 1; oldest multi 1 in DB 1; oldest/newest commit timestamp xid: 0/0; oldest running xid 540; online
    rmgr: Database    len (rec/tot):     42/    42, tx:        540, lsn: 0/01642A40, prev 0/016429C8, desc: CREATE copy dir 1663/1 to 1663/16390
    rmgr: Standby     len (rec/tot):     54/    54, tx:          0, lsn: 0/01642A70, prev 0/01642A40, desc: RUNNING_XACTS nextXid 541 latestCompletedXid 539 oldestRunningXid 540; 1 xacts: 540
    rmgr: XLOG        len (rec/tot):    114/   114, tx:          0, lsn: 0/01642AA8, prev 0/01642A70, desc: CHECKPOINT_ONLINE redo 0/1642A70; tli 1; prev tli 1; fpw true; xid 0:541; oid 24576; multi 1; offset 0; oldest xid 532 in DB 1; oldest multi 1 in DB 1; oldest/newest commit timestamp xid: 0/0; oldest running xid 540; online
    rmgr: Transaction len (rec/tot):     66/    66, tx:        540, lsn: 0/01642B20, prev 0/01642AA8, desc: COMMIT 2021-05-21 15:55:46.363728 EEST; inval msgs: catcache 21; sync
    rmgr: XLOG        len (rec/tot):    114/   114, tx:          0, lsn: 0/01642B68, prev 0/01642B20, desc: CHECKPOINT_SHUTDOWN redo 0/1642B68; tli 1; prev tli 1; fpw true; xid 0:541; oid 24576; multi 1; offset 0; oldest xid 532 in DB 1; oldest multi 1 in DB 1; oldest/newest commit timestamp xid: 0/0; oldest running xid 0; shutdown

The compute node had correctly replayed all the WAL up to the last
record, and opened up. But when you tried to connect to the new
database, the very first requests for the critical relations, like
pg_class, were made with request LSN 0/01642990. That's the last
record that's applicable to a particular block. Because the database
CREATE record didn't bump up the "last written LSN", the getpage
requests were made with too old LSN.

I fixed this by adding a SetLastWrittenLSN() call to the redo of
database CREATE record. It probably wouldn't hurt to also throw in a
call at the end of WAL replay, but let's see if we bump into more
cases like this first.

This doesn't seem to be happening with page server as of 'main'; I was
testing with a version where I had temporarily reverted all the recent
changes to reconstruct control file, checkpoints, relmapper files
etc. from the WAL records in the page server, so that the compute node
was redoing all the WAL. I'm pretty sure we need this fix even with
'main', even though this test case wasn't failing there right now.

knizhnik committed:
[contrib/zenith] [refer #225] if insert WAL position points at the end of WAL page header, then return it back to the page origin

knizhnik committed:
[walproposer] Create replication slot for walproposer to avoid loose of WAL at compute node
+ Check for presence of replication slot

arssher committed:
[walproposer] Fix breaking out of WalProposerPoll and WaitEventSetWait inside.

WAL proposer (as bgw without BGWORKER_BACKEND_DATABASE_CONNECTION) previously
ignored SetLatch, so once caught up it stuck inside WalProposerPoll infinitely.

Futher, WaitEventSetWait didn't have timeout, so we didn't try to reconnect if
all connections are dead as well. Fix that.

Also move break on latch set to the end of the loop to attempt
ReconnectWalKeepers even if latch is constantly set.

Per test_race_conditions (Python version now).

knizhnik committed:
[walproposer] Make it possible to start postgres without reading checkpoint from WAL

+ Check for presence of zenith.signal file to allow skip reading checkpoint record from WAL

+ Pass prev_record_ptr through zenith.signal file to postgres

funbringer committed:
[walredo] Add basic support for Seccomp BPF mode

This patch aims to make our bespoke WAL redo machinery more robust
in the presence of untrusted (in other words, possibly malicious) inputs.

Pageserver delegates complex WAL decoding duties to postgres,
which means that the latter might fall victim to carefully designed
malicious WAL records and start doing harmful things to the system.
To prevent this, it has been decided to limit possible interactions
with the outside world using the Secure Computing BPF mode.

We use this mode to disable all syscalls not in the allowlist.
Please refer to src/backend/postmaster/seccomp.c to learn more
about the pros & cons of the current approach.

+ Fix some bugs in seccomp bpf wrapper

* Use SCMP_ACT_TRAP instead of SCMP_ACT_KILL_PROCESS to receive signals.
* Add a missing variant of select() syscall (thx to @knizhnik).
* Write error messages to an fd stderr's currently pointing to.

ololobus committed:
[walproposer] Simplify WL_LATCH_SET testing in the walproposer

knizhnik committed:
[smgr_api] [contrib/zenith] 1. Do not call mdinit from smgrinit() because it cause memory leak in wal-redo-postgres
2. Add check for local relations to make it possible to use DEBUG_COMPARE_LOCAL mode in SMGR

+ Call smgr_init_standard from smgr_init_zenith

LizardWizzard committed:
[walproposer] [contrib/zenith] support zenith_tenant

this patch adds support for zenith_tenant variable. it has similar
format as zenith_timeline. It is used in callmemaybe query to pass
tenant to pageserver and in ServerInfo structure passed to wal acceptor

lubennikovaav committed:
Implement pg_database_size():
- extend zenith pageserver API to handle new request type;
- add dbsize_hook to intercept db_dir_size() call.

knizhnik committed:
[refer #27] Implement shared relsize cache to improve zenith performance.

Cache relfilenode size returned by zenith_nblocks() and also update it when relation is extended.
Don't update it from zenith_write() or zenith_wallog_page(), since there is no guarantee that these functions wouldn't be called for some page that is not the last one

It can be configured with zenith.relsize_hash_size GUC parameter.
Set it to 0 to disable caching.

arssher committed:
[walproposer] Remove graceful termination of COPY during walproposer recovery.

Rust's postgres_backend currently is too dummy to handle it properly: reading
happens in separate thread which just ignores CopyDone. Instead, writer thread
must get aware of termination and send CommandComplete. Also reading socket must
be transferred back to postgres_backend (or connection terminated completely
after COPY). Let's do that after more basic safkeeper refactoring and right now
cover this up to make tests pass.

ref #388

LizardWizzard committed:
[contrib/zenith] Use authentication token passed as environment variable in connections
to pageserver. Token is passed as cleartext password.

knizhnik committed:
[walproposer] [contrib/zenith] [refer #395] Do no align sart replication position in wal_proppser to segment boundary

lubennikovaav committed:
[test] Add contrib/zenith_test_utils with helpers for testing and debugging.
Now it contains only one function test_consume_xids() for xid wraparound testing.

knizhnik committed:
[walproposer] Change condition for triggering recovery

hlinnaka committed:
[contrib/zenith] Fix race condition while WAL-logging page, leading to CRC errors.

zenith_wallog_page() would call log_newpage() on a buffer, while holding
merely a shared lock on the page. That's not cool, because another backend
could modify the page concurrently. We allow changing hint bits while
holding only a shared lock, and changes on FSM pages, at least. See comments
in XLogSaveBufferForHint() for discussion of this problem.

One instance of the race condition that I was able to capture on my laptop
happened like this:

1. Backend A: needs to evict an FSM page from the buffer cache to make
   room for a new page, and calls zenith_wallog_page() on it. That is
   done while holding a share lock on the page.

2. Backend A: XLogInsertRecord() computes the CRC of the FPI WAL record
   including the FSM page

3. Backend B: Updates the same FSM page while holding only a share lock

4. Backend A: Allocates space in the WAL buffers, and copies the WAL
   record header and the page to the buffers.

At this point, the CRC that backend A computed earlier doesn't match the
contents that were written out to the WAL buffers.

The update of the FSM page in backend B happened from there (fsmpage.c):

	/*
	 * Update the next-target pointer. Note that we do this even if we're only
	 * holding a shared lock, on the grounds that it's better to use a shared
	 * lock and get a garbled next pointer every now and then, than take the
	 * concurrency hit of an exclusive lock.
	 *
	 * Wrap-around is handled at the beginning of this function.
	 */
	fsmpage->fp_next_slot = slot + (advancenext ? 1 : 0);

To fix, make a temporary copy of the page in zenith_wallog_page(), and
WAL-log that. Just like XLogSaveBufferForHint() does.

Fixes https://github.com/zenithdb/zenith/issues/413

sharnoff committed:
[walproposer] Rework walkeeper protocol to use libpq (#60)

The majority of work here is going to be heavily cleaned up soon, but
it's worth giving a brief overview of the changes either way.

* Adds libpqwalproposer, serving a similar function to the existing
  libpqwalreceiver -- to provide access to libpq functions without
  causing problems from directly linking them.

* Adds two new state components, giving (a) the type of libpq-specific
  polling required to move on to the next protocol state and (b) the
  kind of socket events it's waiting on. (These are expected to be
  removed or heavily reworked soon.)

* Changes `WalProposerPoll` to make use of a slightly more specialized
  `AdvancePollState`, which has been completely reworked.

ololobus committed:
Build zenithdb/compute-node:latest in CI (zenithdb/console#125)

lubennikovaav committed:
[walproposer] Skip absent WAL segment removed by pg_resetwal

arssher committed:
Basic safekeeper refactoring and bug fixing.

On the walproposer side,

- Change the voting flow so that acceptor tells his epoch along with giving
  the vote, not before it; otherwise it might get immediately stale. #294
- Adjust to using separate structs for disk and network.

ref #315

knizhnik committed:
Add --sync-safekeepers starting standalone walproposer to sync safekeepers (#439).

It is intended to solve the following problems:

a) Chicken-or-the-egg one: compute postgres needs data directory
   with non-rel files that are downloaded from pageserver by calling
   basebackup@LSN. This LSN is not arbitrary, it must include all
   previously committed transactions and defined through consensus
   voting, which happens... in walproposer, a part of compute node.

b) Just warranting such LSN is not enough, we must also actually commit
   it and make sure there is a safekeeper who knows this LSN is
   committed so WAL before it can be streamed to pageserver -- otherwise
   basebackup will hang waiting for WAL. Advancing commit_lsn without
   playing consensus game is impossible, so speculative 'let's just poll
   safekeepers, learn start LSN of future epoch and run basebackup'
   won't work.

Currently --sync-safekeepers is considered completed when 1) at least majority
of safekeepers and 2) *all* safekeepers with live connection to walproposer
switch to new epoch and advance commit_lsn allowing basebackup to proceed. 2)
limits availablity, but that's because currently we don't have a mechanism
defining which safekeeper should stream WAL into pageserver.

lubennikovaav committed:
zenith_regression_tests.patch
Add alternative output for tablespace test, because tablespaces are not supported in zenith yet

hlinnaka committed:
Add test function to flush the shared buffer cache.

arssher committed:
Rename VCL to epochStartLsn and restart_lsn to truncate_lsn.

epochStartLsn is the LSN since which new proposer writes its WAL in its epoch,
let's be more explicit here.

In several places it also actually meant something we call *commit_lsn* -- the
latest lsn known to be reliably commited (it constantly moves within one wal
proposer).

truncate_lsn is LSN still needed by the most lagging safekeeper. restart_lsn is
terminology from pg_replicaton_slots, but here we don't really have 'restart';
hopefully truncate word makes it clearer.

ololobus committed:
Update Dockerfile

kelvich committed:
Ask pageserver only with LSN's aligned on record boundary.

Now pageserver tracks only last_record_lsn and ignores
last_valids_lsn. We can cause deadlock at start or extreme slowness
during the normal work if we call get_page with LSN of incomplete
record.

Patch by @knizhnik

LizardWizzard committed:
pass tenant id in connection string to safekeeper

sharnoff committed:
Cleanup walproposer changes from #60

Closes #66. Mostly corresponds to cleaning up the states we store. Goes
back to single states for each WalKeeper, and we perform blocking writes
for everything but sending the WAL itself.

A few things have been factored out into libpqwalproposer for
simplicity - like handling the nonblocking status of the connection
(even though it's only changed once).

knizhnik committed:
[refer #506] Correctly initialize all fields of WAL page header for first WAL record of started compute node

arssher committed:
Always advance truncateLsn to commitLsn, keeping it on record boundary.

And take initial value from freshly created slot position. Thus proposer always
starts streaming from the record beginning; it simplifies WAL decoding on
safekeeper.

ololobus committed:
Fix compiler warnings in walproposer.c

arssher committed:
Minor logging editing.

arssher committed:
Fix walproposer starting streaming point.

Send *all* entries (from the beginning, i.e. truncateLsn) to everyone but donor
who doesn't need recovery at all and will receive only new entries. This can be
optimized to avoid sending data which is already persisted (and correct), but
previous such optimization was incorrect.

arssher committed:
Mark all recovery messages as received by the donor.

I forgot to do that in 42316a81d3. Fixes segfault related to attempt to send the
(garbage collected) message second time and queue advancement when donor doesn't
restart.

arssher committed:
Don't FATAL in walproposer when EOF arrives in SS_IDLE state.

arssher committed:
Optimize walproposer starting streaming point.

Safekeepers who are in the same epoch as donor definitely have correct WAL, so
we can send to them since their flushLsn. This required some additionall fuss
due to convention of always starting streaming at the record boundary.

arssher committed:
Simplify a2e929e by storing starting point in walkeeper itself.

hlinnaka committed:
Silence a compiler warning.

This one:

    src/backend/tcop/zenith_wal_redo.c:294:2: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
      294 |  bool enable_seccomp = true;
          |  ^~~~

There are a few more warnings that look a bit more tricky to fix, but let's
at least silence this for now.

In the passing, also move the 'n_synced' local variable closer to where
it's used.

hlinnaka committed:
Remove unused functions for reading non-rel pages.

These could be used to fetch SLRUs and other non-relation things from the
page server. But we don't do that, and have no plans in the near future.

hlinnaka committed:
Misc cleanup in the code that communicates with the page server.

- Remove unused 'system_id' field from ZenithRequest.
- Remove unused 'loaded' variable.
- Remove unused to pack pageserver->client messages, and to unpack
  client->pageserver messages.
- Fix printing the response in debug message (was printing the request
  twice)
- Avoid the overhead of converting request/response to string, unless
  the debug message is really going to be printed
- Formatting fixes.

hlinnaka committed:
Improve the protocol between Postgres and page server.

- Use different message formats for different kinds of response messages.
- Add an Error response message, for passing errors from page server to
  Postgres. An Error response now results in an ereport(ERROR
- Add a flag to requests, to indicate that we actually want the latest
  page version on the timeline, and the LSN is just a hint that we know
  that there haven't been any modifications since that LSN. It is currently
  always set to 'true', but once we start supporting read-only replicas,
  they would set it to false.

This changes the network postgres<->page server protocol, so this needs
corresponding changes in the page server side

Also refactor and fix the zm_to_string() function. The ZenithMessageStr
array was broken, because the array indices didn't match the
ZenithMessageTag enum values.

hlinnaka committed:
Silence compiler warning when building without --enable-seccomp

hlinnaka committed:
Run 'pgindent' on zenith-specific code in contrib/zenith and in walproposer.c

hlinnaka committed:
Silence compiler warning

    contrib/zenith/libpagestore.c: In function ‘zenith_connect’:
    contrib/zenith/libpagestore.c:125:2: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
      125 |  const char **keywords = malloc((noptions + 1) * sizeof(*keywords));
          |  ^~~~~

knizhnik committed:
Implement backpressure for compute node to avoid WAL overflow

hlinnaka committed:
Fix a badly worded comment

sharnoff committed:
Catch walkeeper ErrorResponse in PQgetCopyData

PQgetCopyData can sometimes indicate that the copy is done if the
backend returns an error response. So while we still expect that the
walkeeper never sends CopyDone, we can't expect it to never produce
errors.

hlinnaka committed:
Use buffered I/O for reading commands from stdin.

Whatever the bug mentioned in the FIXME comment was with buffered I/O,
it has been fixed now. This greatly reduces the amount of CPU time spent
in WAL redo.

hlinnaka committed:
Replace fread() with plain read() and a hand-written buffer.

The fread() call required allowing the 'fstat' syscall in the seccomp
configuration, and apparently on some platforms also 'newfstatat', as
Max reported this error:

    Sep 28 15:56:55.522 ERRO wal-redo-postgres: ---------------------------------------
    Sep 28 15:56:55.522 ERRO wal-redo-postgres: seccomp: bad syscall 262
    Sep 28 15:56:55.522 ERRO wal-redo-postgres: ---------------------------------------

I'm afraid of allowing 'newfstatat', that seems like it's opening too
much attack surface, since it allows access to files by filename. Maybe
it's OK, but I'm not sure, but there isn't any fundamental reason why
we'd need to call it, I'm not sure why glibc's fread() wants to call it.
So let's avoid the trouble by writing our own simple buffer over plain
read().

hlinnaka committed:
Store unlogged tables locally, and replace PD_WAL_LOGGED.

The smgr implementation needs to distinguish between unlogged/temp and
regular 'permanent' relations, but the smgr API doesn't currently include
that information. Add a 'relpersistence' field to SmgrRelationData, and
as an argument to smgropen(). However, not all callers of smgropen()
have a relcache entry at hand, so we allow some operations to pass 0,
meaning 'unknown'.

Now that we can store unlogged tables locally, use the same machinery
to handle the buffered GiST and SP-GiST index builds. They populate the
index by inserting all the tuples, and use the shared buffer cache while
they do that. They don't WAL-log the pages while they do that, they log
the whole relation as a separate bulk operation after the build has
finished. That poses a problem for Zenith, where smgrwrite() is a no-op
and we rely on WAL-logging to reconstruct the pages. Solve that problem by
storing the pages locally in the compute node, like an unlogged relation,
until the index build finishes and all the pages have been WAL-logged.
To do that, the smgr needs to know when the caller is an unlogged build
operation like that, so add functions to the Smgr API for that.

With this commit, we no longer generate an FPI record whenever a rel is
extended with an all-zeros page. See github issue #482. That greatly
reduces the amount of WAL generated during bulk loading.

petuhovskiy committed:
Fix queue cleanup in proposer (#93)

Queue was moved further than truncateLsn, when quorumLsn matched end of wal record in the middle of queue message. Fix cleanup of unreceived messages.

Co-authored-by: Arseny Sher <sher-ars@yandex.ru>

hlinnaka committed:
Support read-only nodes

This changes the format of the 'zenith.signal' file. It is now a
human-readable text file, with one line like "PREV LSN: 0/1234568", or
"PREV LSN: none" if the prev LSN is not known, or "PREV LSN: invalid" if
starting up in read-write is not allowed.

Also, if 'zenith.signal' is present, don't try to read the checkpoint
record from the WAL. Trust the copy in pg_control, instead.

ololobus committed:
Initialize FSM/VM pages through buffer cache

To prevent loading them from pageserver.

Author: Konstantin Knizhnik with my extension to VM as well.

arssher committed:
ShutdownConnection instead of ResetConnection in more places.

At least currently risk of busy loop (e.g due to bugs) is much higher than
benefit of additional availability if we immediately reconnect; add interval
between the reconnection attempts.

arssher committed:
Remove a drop of syncSafekeepers complexity.

arssher committed:
Fix compiler warning.

warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  364 |  WalMessage *msgQueueAck = msgQueueHead;

ololobus committed:
Turn off back pressure by default

hlinnaka committed:
Handle partial writes to stdout in WAL redo process.

arssher committed:
Add term history to safekeepers.

See corresponding zenith commit.

arssher committed:
Handle keepalives while receiving WAL in recovery.

Since c310932 safekeeper sometimes sends it.

ref #843

petuhovskiy committed:
Fix truncateLsn update (#101)

truncateLsn is now advanced to `Min(walkeeper[i].feedback.flushLsn)` with taking epochs into account.

lubennikovaav committed:
Clarify the meaning of StandbyReply LSNs used for backpressure

ololobus committed:
[walproposer] Get rid of SAB_Error after rebase

Also see 1632ea43 for details.

lubennikovaav committed:
Use max_replication_apply_lag instead of max_replication_write_lag.
Move backpressure throttling from XlogInsert, to ProcessInterrupts(), to restrict writing operations outside of critical section.

LizardWizzard committed:
Forward pageserver connection string to safekeeper

This is needed for implementation of tenant rebalancing. With this
change safekeeper becomes aware of which pageserver is supposed to be
used for replication from this compute.

This also changes logic of substitution of auth token inside the
connection string. So it is substituted during config variable
parsing and available for both, smgr pageserver connection and
walproposer safekeeper connection.

lubennikovaav committed:
Move backpressure throttling to ProcessInterrupts()

kelvich committed:
Stop building docker images in this repo.

Now docker images are being built in zenith repo as that way we have
sequential version number that allows us to compare compute/storage
versions.

lubennikovaav committed:
Extend replication protocol with ZenithFeedback message.

Add extensible ZenithFeedback part to AppendResponse messages
Pass values sizes together with keys in ZenithFeedback message.

Add standby_status_update fields into ZenithFeedback.
Get rid of diskConsistentLsn field in AppendResponse, because now it is send via ZenithFeedback.
Fix calculation of diskConsistentLsn and instanceSize - take values from latest reply from pageserver

lubennikovaav committed:
Implement cluster size quota for zenith compute node.
Use GUC zenith.max_cluster_size to set the limit.

If limit is reached, extend requests will throw out-of-space error.
When current size is too close to the limit - throw a warning.

Do not apply size quota to autovacuum process

Add pg_cluster_size() funciton in zenith extension

petuhovskiy committed:
[walproposer] Async WAL append (#105)

Implement async wp <-> sk protocol, send WAL messages ahead of feedback replies.

New SS_ACTIVE state is introduced instead of former SS_SEND_WAL / SS_SEND_WAL_FLUSH / SS_RECV_FEEDBACK.

lubennikovaav committed:
Fix walsender to work with zenith style standbyReply that sends non-zero flushLsn.
Clean up backpressure defaults.

knizhnik committed:
Prevent recursive call of XLogBeginInsert

refer #1015

ololobus committed:
Do not copy the obsolete apply_conf binary into Docker image

knizhnik committed:
Report back-pressure trottling status of backend

knizhnik committed:
Add max_replication_write_lag

knizhnik committed:
Use local relation cache for smgr_exists

refer  #1077

knizhnik committed:
Use local relation cache for smgr_exists

refer  #1077

petuhovskiy committed:
Reorder walproposer code in a more natural order (#112)

Now functions in walproposer.c go in chronological order

petuhovskiy committed:
Simplify walproposer code (#114)

* Clean up walproposer states

* Migrate AsyncReadFixed to AsyncReadMessage

* Handle flushWrite better a bit

* Update SS_ACTIVE event set in single place

Now event set is updated only in the end of HandleActiveState, after
all handlers code was executed.

* Add comment on SS_ACTIVE write event

* Add TODO for SS_ACTIVE DesiredEvents

petuhovskiy committed:
Don't change propTerm after quorum is acquired (#107)

petuhovskiy committed:
walproposer renames (#116)

* Rename walkeeper to safekeeper

* Rename message variables as request/response

knizhnik committed:
Do not throttle wal sender

knizhnik committed:
Do no throttle wal sender

petuhovskiy committed:
Don't track acks in walproposer (#119)

hlinnaka committed:
Silence excessively noisy logging from walproposer.

In the passing, switch a few places to ereport() instead of elog(), to
avoid the overhead of constructing the string when it's not logged.

Fixes https://github.com/zenithdb/zenith/issues/1066

ololobus committed:
Calculate pages checksum in the WAL redo loop

antons-antons commented on src/backend/tcop/zenith_wal_redo.c:
```diff
@@ -683,6 +683,9 @@ GetPage(StringInfo input_message)
 	/* single thread, so don't bother locking the page */
 
+	// Calculate checksums on the newly materialized page if needed.
+	PageSetChecksumInplace(page, blknum);
```
If we want to be pedantic we should validate checksum before running redo (in PushPage)
also, Log record CRC should be validated as well, currently it's only done in XLogReadRecord but not DecodeXLogRecord (used directly in ApplyRecord)

ololobus replied:
> If we want to be pedantic we should validate checksum before running redo (in PushPage)

IIUC, checksum can be not up-to-date in the full-page-image inside WAL record, as Postgres makes the final update on them only on eviction. That's the reason actually, why I stuck with these PRs. On pageserver we take FPIs from WAL records and store them inside timeline data as 'pages', not original WAL records. So next time we pass these pages back to the Postgres, it complains that they have a wrong CRC. Yet, if I force WAL redo for them, everything is fine, as we stamp a proper CRC on them in the GetPage()

> also, Log record CRC should be validated as well, currently it's only done in XLogReadRecord but not DecodeXLogRecord (used directly in ApplyRecord)

Good catch, thanks! I'll have a look on this part too

knizhnik committed:
Add --sysid parameter to initdb

petuhovskiy committed:
Allow to join empty safekeeper to existing cluster (#123)

knizhnik committed:
Add utilities for recovery database fmro WAL

knizhnik committed:
Calculate wal position in restore_from_wal scripts

refer #1169

arssher committed:
Reduce walproposer logging after ca5e7beaf.

arssher committed:
Revert "Use local relation cache for smgr_exists"

This reverts commit 45dd8911ec13fd47882685f6d81d7b73696e1b84.

It introduced stable test_isolation failure. There was an idea that adding
strict backpressure settings would help, as absense of this commit could behave
as natural backpressure, but that didn't help. No better fix is immediately
available, so let's revert until sorting this out.

ref https://github.com/zenithdb/zenith/issues/1238
ref https://github.com/zenithdb/zenith/pull/1239

hlinnaka committed:
Change the unit of cluster size limit GUC to MB, and other fixes.

The GUC is a 32-bit integer, so if the base unit is bytes, the max
limit you can set is only 2 GB. Furthermore, the web console assumed
that the unit is in MB, and set it to 10000 meaning 10 GB, but in
reality it was set to just 10 kB.

Remove the WARNINGs related to cluster size limit. That was probably
supposed to be DEBUG5 or something, because it's extremely noisy
currently. You get the WARNING for *every block* when a relation is
extended.

Some kind of a WARNING when you approach the limit would make sense,
but it's difficult to do in a sensible way with WARNINGs from the
server. Firstly, most applications will ignore WARNINGs, in which case
they don't accomplish anything. If an application forwards them to the
user, that's not great either unless the application user happens to
be the DBA. If you're lucky, the WARNINGs end up in an application log
and the DBA is alerted, but printing the message for every relation
extension is too noisy for that too. An email alert would probably be
best, outside Postgres.

Also don't enforce the limit when extending a temporary or unlogged
relation. They don't count towards the cluster size limit, so it seems
weird to error out on them. And reword the error message a bit.

Fixes https://github.com/zenithdb/zenith/issues/1233

hlinnaka committed:
Improve error handling while connecting to page server.

If anything goes wrong while establishing a connection, don't leak the
socket.

Also, if you get an error while sending the GetPage request, kill the
connection. It's not clear what state it's in, so better to reconnect.

hlinnaka committed:
Make getpage requests interruptible.

Fixes https://github.com/zenithdb/zenith/issues/1224

hlinnaka committed:
Fix memory leak of messages received from safekeepers.

Fixes https://github.com/zenithdb/zenith/issues/822

knizhnik committed:
Initialize pgxactoff for walproposer

refer #1244

arssher committed:
Fix compilation warning after 8524105dc.

arssher committed:
Fix more compiler warnings.

knizhnik committed:
Update src/backend/access/gin/ginfast.c

Co-authored-by: Heikki Linnakangas <heikki.linnakangas@iki.fi>

hlinnaka committed:
Remove dead code in handling ZenithFeedback part of an AppendResponse.

The constructed StringInfoData 'z' variable wasn't used for anything, we
passed the original 's' StringInfo directly to ParseZenithFeedbackMessage.
That's fine, but let's remove the dead code.

lubennikovaav committed:
Add backpressure_lsns() function. Fix zenith feedback processing

antons-antons reviewed (approved):
Yup, it looks like the right amount of crc validation to protect from data corruption on storage and in transit

antons-antons committed:
Expose reading a relation page at a specific LSN (#131)

* Expose reading a relation page at a specific LSN

* Addressing comments

knizhnik committed:
Add warning fr unrecgonized GUCs with zenith prefix

refer #1262

ololobus commented:
> So that forces crc checks in wal-redo process. Will compute check them too after the getpage() request?

Postgres uses storage manager and our zenith_read() where GetPage@LSN is called. After that page goes into buffer manager, where checksum is verified (ReadBufferExtended->ReadBuffer_common->PageIsVerifiedExtended). So this path doesn't differ from vanilla, we just have to tell Postgres to verify checksums (during initdb phase).

kelvich committed:
Fix zenith_test_utils linkage on macOS

Use function pointer to perform a cross-extension calls.

petuhovskiy committed:
Count WAL flushes in walreceiver (#139)

hlinnaka committed:
Populate relsize cache when relation is created.

Postgres can perform an smgrnblocks() call on the relation right after
creating it, and we don't update the last-written LSN on smgrcreate().

Perhaps we should update last-written LSN, instead. This isn't
bulletproof.

knizhnik committed:
Remove absolute pathes from restore_from_wal scripts

knizhnik committed:
Remove absolute pathes from restore_from_wal scripts

knizhnik committed:
Update scripts for recovery from WAL

knizhnik committed:
Make it possible to specify port and datadir for restore_from_wal

knizhnik committed:
Move recovery scripts to main repo

hlinnaka committed:
Fix pg_table_size() on a view

arssher committed:
Don't set commitLsn to truncateLsn.

It might jump back (on compute) this way, which is not fatal but violates sanity
checks.

petuhovskiy committed:
Raise fatal error on failed recovery (#147)

antons-antons committed:
Enable dumping corrupt WAL segments (#145)

* Enable dumping corrupt WAL segments

 Add ability to dump WAL segment with corrupt page headers and recrods
 skips over missing/broken page headers
 skips over misformatted log recrods
 allows dumping log record from a particular file starting from an
optional offset
 (without a need of carefully crafted input)

petuhovskiy committed:
Don't hold walproposer WAL in memory (#141)

WAL is no longer in memory to prevent OOM in the compute. Removed in-memory queue because it's not needed anymore. When streaming, WAL is now read directly from disk. Every safekeeper has a separate XLogReader. walproposer will now read as much WAL as it can for a single AppendRequest message, it can help with recovering lagging safekeepers. Because Recovery needs to save WAL for streaming, now walproposer can write WAL to disk and `--sync-safekeepers` mode will create pg_wal directory if needed. Replication slot `restart_lsn` is now synced with `truncate_lsn` to prevent truncation of disk WAL until needed.

arssher committed:
Give up connection attempt to safekeeper after timeout.

Enforces reconnection soon when packets are dropped, e.g. after turning ec2
instance off.

ref https://github.com/neondatabase/neon/issues/1491

lubennikovaav committed:
Rename 'wal_acceptors' GUC to 'safekeepers'

knizhnik committed:
Avoid redundand memory allocation and sycnhronization in walredo (#144)

* Avoid redundand memory allocation and sycnhronization in walredo

* Address review comments

* Reduce number of temp buffers and size of inmem file storage for wal redo postgres

* Misc cleanup

Add comments on 'inmem_smgr.c', remove superfluous copy-pasted comments,
pgindent.

Co-authored-by: Heikki Linnakangas <heikki.linnakangas@iki.fi>

knizhnik committed:
Fix missed include for InRecovery (#149)

* Fix missed include for InRecovery

* Fix missed include for InRecovery (used only in debug version with --enable--cassert)

hlinnaka committed:
Avoid "bad syscall 39" on assertion failure in WAL redo process.

ExceptionalCondition calls getpid(), which is currently forbidden by
seccomp. You only get there if something else went wrong, but the "bad
syscall" error hides the underlying cause of the error, which makes
debugging hard.

hlinnaka committed:
Add error context, if replaying a WAL record fails in WAL redo process.

hlinnaka committed:
Add WARNING for debugging purposes.

hlinnaka committed:
Turn Assertion into elog(ERROR), to help with debugging.

This error is happening in the 'pg_regress' test in the CI, but not on
my laptop. Turn it into an ERROR, so that we get the error context and
backtrace of it.

hlinnaka committed:
Fix errors in WAL redo about relpersistence mismatch.

In the WAL redo process, even "permanent" buffers are stored in the
local buffer cache. Need to pass RELPERSISTENCE_PERMANENT to smgropen()
in that case.

arssher committed:
Don't log 'last written LSN ahead of flushed'.

That's a valid case, as edited comment says.

https://github.com/neondatabase/neon/issues/1303

knizhnik committed:
Perform inmem_smgr cleaup after processing each record (#154)

* Perform inmem_smgr cleaup after processing each record

* Prevent eviction of wal redo target page

* Prevent eviction of wal redo target page frmo temp buffers

arssher committed:
Send timeline_start_lsn in Elected and receive it in VoteResponse messages.

To support remembering it on safekeeper. Currently compute doesn't know initial
LSN on non-first boot (though it could get it from pageserver in theory), so we
rely on safekeepers to fetch it back.

While changing the protocol, also add node_id to AcceptorProposerGreeting.

arssher committed:
Verify basebackup LSN against consensus LSN in walproposer.

If not, such basebackup (clog etc) is inconsistent and must be retaken.

Basebackup LSN is taken by exposing xlog.c RedoStartLSN in shmem.

ref https://github.com/neondatabase/neon/issues/594

hlinnaka committed:
Avoid extending relation in the WAL redo process.

It's a waste of time, and otherwise you can run into the MAX_PAGES limit.

Fixes https://github.com/neondatabase/neon/issues/1615

arssher committed:
Shut down instance on basebackup LSN mismatch.

To force making basebackup again.

ololobus committed:
Use compute-tools from the new neondatabase Docker Hub repo

yeputons committed:
zenith_test_utils extension: add neon_xlogflush()

This function is to simplify complex WAL generation in https://github.com/neondatabase/neon/pull/1574

`pg_logical_emit_message` is the easiest way to get a big WAL record, but:
* If it's transactional, it gets `COMMIT` record right after
* If it's not, WAL is not flushed at all. The function helps here, so we
  don't rely on the background WAL writer.

I suspect the plain `xlogflush()` name may collide in the future, hence the prefix.

hlinnaka committed:
Reduce noise in the logs from inmem_write()

I'm seeing a lot of these warnings from B-tree SPLIT records:

    WARNING:  inmem_write() called for 1663/12990/16397.0 blk 2630: used_pages 0
    CONTEXT:  WAL redo at 1/235A1B50 for Btree/SPLIT_R: level 0, firstrightoff 368, newitemoff 408, postingoff 0

That seems OK, replaying a split record legitimately accesses many buffers:
the left half, the right half, left sibling, right sibling, and child.

We could bump up 'temp_buffers' (currently 4), but I didn't do that
beceause it's also good to get some test coverage for the
inmem_smgr.c.

hlinnaka committed:
Misc cleanup in libpagestore.c.

- Fix typos
- Change Zenith -> Neon in the ZENITH_SMGR tag that's printed in error
  messages that is user-visible, and in various function names and comments
  that are not user-visible.
- pgindent
- Remove comment about zm_to_string() leaking memory. It doesn't.
- Re-word some error messages to match PostgreSQL error message style guide

ololobus committed:
Do not force FPI if data checksums are enabled

ololobus committed:
Use compute_ctl instead of zenith_ctl (#162)

hlinnaka committed:
Improve error messages on seccomp loading errors.

At https://github.com/neondatabase/neon/pull/1783#issuecomment-1136144433,
Kirill saw case where the WAL redo process failed to open /dev/null.
That's pretty weird, and I have no idea what might be causing it, but
with this patch we'll at least get a little more details if it happens
again. This will print the OS error (with %m) if it happens, and also
distinguishes between the two error cases that previously both emitted
the 'failed to open a test file' error.

lubennikovaav committed:
Rename contrib/zenith to contrib/neon. Rename custom GUCs:
- zenith.page_server_connstring -> neon.pageserver_connstring
- zenith.zenith_tenant -> neon.tenant_id
- zenith.zenith_timeline -> neon.timeline_id
- zenith.max_cluster_size -> neon.max_cluster_size

hlinnaka committed:
Address Matthias' review comments

arssher committed:
Fix basebackup LSN comparison in walproposer.

as basebackup LSN always skips over page header

ololobus committed:
Allow specifiyng the different compute-tools tag (#170)

Part of https://github.com/neondatabase/neon/pull/1838

knizhnik committed:
Do not allocate shared memory for wal_redo process (#165)

* Do not allocate shared memory for wal_redo process

* Add comment

knizhnik committed:
Add check for NULL for malloc in InternalIpcMemoryCreate (#173)

* Add check for NULL for malloc in InternalIpcMemoryCreate

* apply pgindent

SomeoneToIgnore committed:
Rename ZenithFeedback (#174)

SomeoneToIgnore committed:
Remove callmemaybe bits from compute (#172)

petuhovskiy committed:
Panic on receiving higher term (#176)

hlinnaka committed:
Don't print JWT token to log
